### DISCUSSION TITLE
Is there a way to access source locations from a plugin?

### DISCUSSION THREAD

**Discussion by vimtor** [posthtml/posthtml#359] [Q&A] [answered]
Hi there!

I was wondering if there was a way to access source locations from a plugin. I need to add a `data-attribute` with the file name, column and line to each HTML tag.

I've seen in the [posthtml-parser](https://github.com/posthtml/posthtml-parser#sourcelocations) that an options `sourLocations` is available, but I cannot get the source location anyway.

Thanks!

**Comment by Scrum** [comment, marked as answer]
Hi yes this is available when using posthtml since version 0.16.0

Here is an example that implements what you need through a plugin, referring to the properties of the tree and the node.

```js
const fs = require('fs');
const path = require('path');

const posthtml = require('posthtml');

const files = [
  'files/index.html',
  'files/test.html'
]

const plugin = function (tree) {
  tree.walk(node => {
    Object.assign(node, { 
      attrs: { 
        'data-file-name': tree.options.from,
        'data-location': JSON.stringify(node.location)
      } 
    });
    
    return node;
  })
  return tree;
}


files.forEach(file => {
  const html = fs.readFileSync(path.join(__dirname, file), 'utf-8');

  const result = posthtml(plugin)
    .process(html, { 
      sync: true,
      from: file,
      sourceLocations: true
    })
    .html

  console.log(result)
});
```

> Please note that in this example you need to forward the value yourself `from`. You can also use it [posthtml-cli](https://github.com/posthtml/posthtml-cli) which automatically sets the value `from` when processing files.
